lib/handler/mruby/channel.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2017 Ritta Narita
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <mruby.h>
+#include <mruby/array.h>
+#include <mruby/error.h>
+#include <mruby/hash.h>
+#include <mruby/string.h>
+#include <mruby/class.h>
+#include <mruby/variable.h>
+#include "h2o/mruby_.h"
+#include "embedded.c.h"
+
+struct st_h2o_mruby_channel_context_t {
+    h2o_mruby_context_t *ctx;
+    mrb_value receiver;
+};
+
+static void attach_receiver(struct st_h2o_mruby_channel_context_t *ctx, mrb_value receiver)
+{
+    assert(mrb_nil_p(ctx->receiver));
+    ctx->receiver = receiver;
+    mrb_gc_register(ctx->ctx->shared->mrb, receiver);
+}
+
+static mrb_value detach_receiver(struct st_h2o_mruby_channel_context_t *ctx)
+{
+    mrb_value ret = ctx->receiver;
+    assert(!mrb_nil_p(ret));
+    ctx->receiver = mrb_nil_value();
+    mrb_gc_unregister(ctx->ctx->shared->mrb, ret);
+    mrb_gc_protect(ctx->ctx->shared->mrb, ret);
+    return ret;
+}
+
+static void on_gc_dispose_channel(mrb_state *mrb, void *_ctx)
+{
+    struct st_h2o_mruby_channel_context_t *ctx = _ctx;
+    assert(ctx != NULL); /* ctx can only be disposed by gc, so data binding has been never removed */
+    mrb_gc_unregister(mrb, ctx->receiver);
```

Comment:
Could you add a check that ensures ctx->receiver is not nil? i.e. if (!mrb_nil_p(ctx->receiver))

Reply:
I fixed to check if `ctx->receiver` is not nil before `mrb_gc_unregister `.
https://github.com/h2o/h2o/pull/1336/commits/4312b539294bbb81d7cdf1621a05a08e53c2e0dd

Reply:
At this time we definitely set `receivers` in initialization, so this nil check you newly added is no longer needed. Sorry for the confusion 😵

Reply:
Sorry, I forgot to check it. I deleted it.
https://github.com/h2o/h2o/pull/1336/commits/03c0a42d76750c3cd95f1eab894be9151ac53b5e